system/Filters/Filters.php

```diff
@@ -173,55 +174,129 @@ public function run(string $uri, string $position = 'before')
-                    // short circuit - bypass any other filters
-                    return $result;
-                }
-                // Ignore an empty result
-                if (empty($result)) {
-                    continue;
-                }
+                continue;
+            }
 
+            // If the response object was sent back,
+            // then send it and quit.
+            if ($result instanceof ResponseInterface) {
+                // short circuit - bypass any other filters
                 return $result;
             }
 
-            if ($position === 'after') {
-                $result = $class->after(
-                    $this->request,
-                    $this->response,
-                    $this->argumentsClass[$className] ?? null
-                );
+            // Ignore an empty result
+            if (empty($result)) {
+                continue;
+            }
 
-                if ($result instanceof ResponseInterface) {
-                    $this->response = $result;
+            return $result;
+        }
 
-                    continue;
-                }
+        return $this->request;
+    }
+
+    private function runAfter(array $filterClasses): ResponseInterface
+    {
+        foreach ($filterClasses as $className) {
+            $class = new $className();
+
+            if (! $class instanceof FilterInterface) {
+                throw FilterException::forIncorrectInterface(get_class($class));
             }
+
+            $result = $class->after(
+                $this->request,
+                $this->response,
+                $this->argumentsClass[$className] ?? null
+            );
+
+            if ($result instanceof ResponseInterface) {
+                $this->response = $result;
+
+                continue;
+            }
+        }
+
+        return $this->response;
+    }
+
+    /**
+     * Runs required filters for the specified position.
+     *
+     * @return RequestInterface|ResponseInterface|string|null
+     *
+     * @phpstan-param 'before'|'after' $position
+     *
+     * @throws FilterException
+     */
+    public function runRequired(string $position = 'before')
+    {
+        if (! isset($this->config->required[$position]) || $this->config->required[$position] === []) {
+            return $position === 'before' ? $this->request : $this->response;
+        }
+
+        // Set the toolbar filter to the last position to be executed
+        if (
+            in_array('toolbar', $this->config->required['after'], true)
+            && ($count = count($this->config->required['after'])) > 1
+            && $this->config->required['after'][$count - 1] !== 'toolbar'
+        ) {
+            array_splice(
+                $this->config->required['after'],
+                array_search('toolbar', $this->config->required['after'], true),
+                1
+            );
+            $this->config->required['after'][] = 'toolbar';
+        }
```

Comment:
A single iteration may be more efficient and easier to read:
```suggestion
        $afters = [];
        $found = true;
        foreach ($this->config->required['after'] as $alias) {
            if ($alias === 'toolbar') {
                $found = true;
                continue;
            }

            $afters[] = $alias;
        }

        if ($found) {
            $afters[] = 'toolbar';
        }
        $this->config->required['after'] = $afters;
```